llvm/lib/Target/X86/X86ISelDAGToDAG.cpp

```diff
@@ -1036,6 +1036,42 @@ void X86DAGToDAGISel::PreprocessISelDAG() {
 
       break;
     }
+    case ISD::LOAD: {
+      // If this is a XMM/YMM load of the same lower bits as another YMM/ZMM
+      // load, then just extract the lower subvector and avoid the second load.
+      auto *Ld = cast<LoadSDNode>(N);
+      MVT VT = N->getSimpleValueType(0);
+      if (ISD::isNormalLoad(Ld) && Ld->isSimple() &&
```

Comment:
Should we check `!isVolatile` instead of `isSimple`?

Reply:
I suggest early bail out here by reversing the condition.

Reply:
If we do that then we'd probably need to ensure `Ld->isAtomic() == UserLd->isAtomic()` - would that be acceptable?

Reply:
Good point. Then `isSimple` sounds good to me.